conda/gateways/repodata/__init__.py

```diff
@@ -0,0 +1,283 @@
+            resp = session.get(url, headers=headers, proxies=session.proxies, timeout=timeout)
+            if self._log.isEnabledFor(logging.DEBUG):
+                self._log.debug(stringify(resp, content_max_len=256))
+            resp.raise_for_status()
+
+        if resp.status_code == 304:
+            raise Response304ContentUnchanged()
+
+        def maybe_decompress(filename, resp_content):
+            return ensure_text_type(
+                bz2.decompress(resp_content) if filename.endswith(".bz2") else resp_content
+            ).strip()
+
+        json_str = maybe_decompress(filename, resp.content)
+
+        # We explictly no longer add these tags to the large `resp.content` json
+        saved_fields = {"_url": self._url}
+        _add_http_value_to_dict(resp, "Etag", saved_fields, "_etag")
+        _add_http_value_to_dict(resp, "Last-Modified", saved_fields, "_mod")
+        _add_http_value_to_dict(resp, "Cache-Control", saved_fields, "_cache_control")
+
+        state.clear()
+        state.update(saved_fields)
+
+        return json_str
+
+def _add_http_value_to_dict(resp, http_key, d, dict_key):
+    value = resp.headers.get(http_key)
+    if value:
+        d[dict_key] = value
+
+
+@contextmanager
+def conda_http_errors(url, repodata_fn):
+    """
+    Use in a with: statement to translate requests exceptions to conda ones.
+    """
+    try:
+        yield
+    except RequestsProxyError:
+        raise ProxyError()  # see #3962
+
+    except InvalidSchema as e:
+        if "SOCKS" in str(e):
+            message = """\
+Requests has identified that your current working environment is configured
+to use a SOCKS proxy, but pysocks is not installed.  To proceed, remove your
+proxy configuration, run `conda install pysocks`, and then you can re-enable
+your proxy configuration.
+"""
+            raise CondaDependencyError(message)
+        else:
+            raise
+
+    except SSLError as e:
+        # SSLError: either an invalid certificate or OpenSSL is unavailable
+        try:
+            import ssl  # noqa: F401
+        except ImportError:
+            raise CondaSSLError(
+                f"""\
+OpenSSL appears to be unavailable on this machine. OpenSSL is required to
+download and install packages.
+
+Exception: {e}
+"""
+            )
+        else:
+            raise CondaSSLError(
+                f"""\
+Encountered an SSL error. Most likely a certificate verification issue.
+
+Exception: {e}
+"""
+            )
+
+    except (ConnectionError, HTTPError) as e:
+        status_code = getattr(e.response, "status_code", None)
+        if status_code in (403, 404):
+            if not url.endswith("/noarch"):
+                log.info(
+                    "Unable to retrieve repodata (response: %d) for %s",
+                    status_code,
+                    url + "/" + repodata_fn,
+                )
+                raise RepodataIsEmpty(
```

Comment:
the old one did "return None" here, which we can't do in the with handler.

Reply:
hidden in here is a simple "retry if current_repodata.json" is not available behavior.